# Домашнее задание: Кросс-валидация, GridSearchCV и SVM

## Цель работы

Применить различные методы кросс-валидации и освоить процесс подбора гиперпараметров для SVM. Сравнить качество моделей Logistic Regression и SVM на выбранном датасете.

## Описание задания

Выполнить полный цикл анализа данных с использованием техник кросс-валидации, подбора параметров и сравнения алгоритмов. Необходимо воспроизвести все методологии, рассмотренные на уроке, адаптировав их под особенности выбранного датасета.

## Датасеты на выбор

1. **Датасет 1**: Wine (использовать load_wine из sklearn.datasets)
2. **Датасет 2**: Digits (использовать load_digits из sklearn.datasets, но только классы 0 и 1 для бинарной классификации)

## Этапы работы и требования

### 1. Загрузка и первичный анализ данных

- Загрузите выбранный датасет
- Проведите первичный анализ: размерность, распределение классов, наличие пропусков
- Визуализируйте распределение классов
- Проверьте масштабы признаков

### 2. Подготовка данных

- Разделите данные на обучающую и тестовую выборки (80/20)
- Используйте stratified split для сохранения пропорций классов
- Проведите стандартизацию признаков (StandardScaler)
- Убедитесь, что стандартизация применена правильно (fit только на train)

### 3. Сравнение методов кросс-валидации

Обучите Logistic Regression с помощью разных методов кросс-валидации и сравните результаты:

- **K-Fold CV (K=5)**: вычислите scores
- **Stratified K-Fold CV (K=5)**: вычислите scores
- **Leave-One-Out CV** (на подмножестве из 100 примеров): вычислите accuracy
- **Nested CV**: используя GridSearchCV внутри для подбора C из [0.1, 1, 10, 100]

Для каждого метода выведите:
- Среднее значение метрики
- Стандартное отклонение
- Минимум и максимум
- Таблицу сравнения всех методов

Вывод: какой метод дал наиболее надежную оценку и почему?

### 4. Базовая модель SVM

- Обучите SVM с linear kernel и параметром C=1.0
- Вычислите train и test accuracy
- Определите количество опорных векторов
- Сравните с базовой Logistic Regression

### 5. Исследование влияния параметров SVM

**Часть A: RBF ядро vs Linear ядро**
- Обучите SVM с RBF kernel (gamma='scale', C=1.0)
- Сравните с linear kernel
- Оцените train и test accuracy
- Визуализируйте количество опорных векторов

**Часть B: Влияние параметра C**
- Протестируйте C = [0.01, 0.1, 1, 10, 100] с RBF kernel
- Для каждого C вычислите train и test accuracy
- Постройте график зависимости accuracy от C
- Определите оптимальное C визуально
- Объясните что происходит при малых и больших C

**Часть C: Влияние параметра Gamma**
- Протестируйте gamma = [0.0001, 0.001, 0.01, 0.1, 1] с C=1.0
- Для каждого gamma вычислите train и test accuracy
- Постройте график зависимости accuracy от gamma
- Определите оптимальное gamma визуально
- Объясните эффект gamma на сложность модели

### 6. GridSearchCV для SVM

- Создайте сетку параметров:
  - kernel: ['linear', 'rbf']
  - C: [0.1, 1, 10, 100]
  - gamma: [0.001, 0.01, 0.1, 1] (только для RBF)
- Выполните GridSearchCV с StratifiedKFold(n_splits=5)
- Выведите top-10 комбинаций по CV score
- Обучите финальную модель на лучших параметрах
- Вычислите accuracy на обучающей и тестовой выборках
- Проанализируйте разницу между best CV score и test score

### 7. Nested Cross-Validation

**Для Logistic Regression:**
- Outer CV: StratifiedKFold(n_splits=5)
- Inner CV: StratifiedKFold(n_splits=3)
- Параметры для подбора: C из [0.1, 1, 10, 100]
- Вычислите итоговую среднюю accuracy и std

**Для SVM:**
- Outer CV: StratifiedKFold(n_splits=5)
- Inner CV: StratifiedKFold(n_splits=3)
- Параметры для подбора: C и gamma (сокращенная сетка)
- Вычислите итоговую среднюю accuracy и std

Сравните результаты Nested CV для обеих моделей

### 8. Итоговое сравнение моделей

Создайте таблицу сравнения для оптимальных моделей LR и SVM:

| Метрика | LR | SVM | Победитель |
|---------|----|----|--------|
| Test Accuracy | | | |
| Nested CV Mean | | | |
| Nested CV Std | | | |
| Время обучения | | | |
| Интерпретируемость | | | |
| Использованные опорные векторы | | | |

Вывод: какая модель лучше и почему?

## Требования к коду

- Весь код должен быть хорошо организован и прокомментирован
- Используйте информативные названия переменных
- Выводите результаты в виде таблиц (pandas DataFrame)
- Строите необходимые графики для визуализации
- Используйте print() для подробного вывода информации
- Устанавливайте random_state=42 для воспроизводимости

## Требования к визуализации

- График распределения классов
- График зависимости accuracy от параметра C (шаг 5B)
- График зависимости accuracy от параметра gamma (шаг 5C)
- Таблицы сравнения методов CV
- Таблица top-10 комбинаций из GridSearchCV
- Сравнительная таблица финальных моделей

## Формат отчета

По результатам работы необходимо предоставить Jupyter Notebook, содержащий:

1. Весь программный код с комментариями
2. Все таблицы сравнения результатов
3. Все графики

После кода добавьте текстовый отчет с выводами:

1. Анализ данных: характеристики датасета и его особенности
2. Сравнение методов кросс-валидации: какой наиболее надежный и почему?
3. Влияние параметров SVM: как C и gamma влияют на качество модели?
4. Результаты GridSearchCV: найденные оптимальные параметры и их эффект
5. Результаты Nested CV: честная оценка качества моделей
6. Финальное сравнение LR и SVM: преимущества и недостатки каждого подхода
7. Рекомендация: какую модель использовать для этого датасета и почему?

Отчет должен быть структурированным и содержать цифры из полученных результатов.

## Выбор датасета

Выберите один из трех предложенных датасетов и раскомментируйте соответствующий код

In [ ]:
from sklearn import datasets
from sklearn.model_selection import (
    train_test_split, KFold, StratifiedKFold, LeaveOneOut,
    cross_val_score, GridSearchCV
)
from sklearn.preprocessing import StandardScaler
from sklearn.linear_model import LogisticRegression
from sklearn.svm import SVC
from sklearn.metrics import accuracy_score, confusion_matrix, classification_report
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import time

np.random.seed(42)

In [ ]:
data = datasets.load_wine()
X, y = data.data, data.target
dataset_name = "Wine"
y = (y == 0).astype(int)

# ДАТАСЕТ 2: Digits (классы 0 и 1)
# data = datasets.load_digits()
# mask = (data.target == 0) | (data.target == 1)
# X = data.data[mask]
# y = data.target[mask]
# dataset_name = "Digits (0 vs 1)"

print(f"Датасет: {dataset_name}")
print(f"Размер: {X.shape}")
print(f"Классы: {np.unique(y)}")

Датасет: Wine
Размер: (178, 13)
Классы: [0 1]


---

## ШАГ 1-2: Анализ данных и подготовка

In [ ]:
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=0.2, stratify=y, random_state=42)

print("Размер train:", X_train.shape)
print("Размер test:", X_test.shape)

scaler = StandardScaler()

X_train_scaled = scaler.fit_transform(X_train)
X_test_scaled = scaler.transform(X_test)

print("Среднее по train:")
print(X_train_scaled.mean(axis=0)[:5])

print("\nStd по train:")
print(X_train_scaled.std(axis=0)[:5])



Размер train: (142, 13)
Размер test: (36, 13)
Среднее по train:
[ 4.35957999e-15  1.10475009e-15  2.02576610e-15  2.13287916e-15
 -3.12738880e-18]

Std по train:
[1. 1. 1. 1. 1.]


---

## ШАГ 3: Сравнение методов кросс-валидации

In [ ]:
log_reg = LogisticRegression(solver="liblinear", max_iter=1000, random_state=42)

kf = KFold(n_splits=5, shuffle=True, random_state=42)
kf_scores = cross_val_score(log_reg, X_train_scaled, y_train, cv=kf, scoring="accuracy")

skf = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
skf_scores = cross_val_score(log_reg, X_train_scaled, y_train, cv=skf, scoring="accuracy")

X_subset = X_train_scaled[:100]
y_subset = y_train[:100]
loo = LeaveOneOut()
loo_scores = cross_val_score(log_reg, X_subset, y_subset, cv=loo, scoring="accuracy")

param_grid = {
    "C": [0.1, 1, 10, 100]
}

inner_cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
outer_cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

grid_search = GridSearchCV(
    estimator=log_reg,
    param_grid=param_grid,
    cv=inner_cv,
    scoring="accuracy"
)

nested_scores = cross_val_score(
    grid_search,
    X_train_scaled,
    y_train,
    cv=outer_cv,
    scoring="accuracy"
)

print("K-Fold CV scores:", kf_scores)
print("Средняя accuracy:", kf_scores.mean())

print("Stratified K-Fold CV scores:", skf_scores)
print("Средняя accuracy:", skf_scores.mean())

print("LOO accuracy (subset of 100):", loo_scores.mean())

print("Nested CV scores:", nested_scores)
print("Средняя accuracy (Nested CV):", nested_scores.mean())



Nested CV scores: [1.         0.96551724 0.92857143 1.         1.        ]
Средняя accuracy (Nested CV): 0.9788177339901478


---

## ШАГ 4: Базовая модель SVM

In [ ]:
# ВАШ КОД ЗДЕСЬ
# Обучите SVM с linear kernel
# Сравните с Logistic Regression

---

## ШАГ 5: Исследование влияния параметров SVM

In [ ]:
svm_linear = SVC(
    kernel="linear",
    C=1.0,
    random_state=42
)

svm_linear.fit(X_train_scaled, y_train)

y_train_pred_svm = svm_linear.predict(X_train_scaled)
y_test_pred_svm = svm_linear.predict(X_test_scaled)

svm_train_acc = accuracy_score(y_train, y_train_pred_svm)
svm_test_acc = accuracy_score(y_test, y_test_pred_svm)

n_support_vectors = svm_linear.n_support_
total_support_vectors = svm_linear.support_vectors_.shape[0]

log_reg = LogisticRegression(
    C=1.0,
    solver="liblinear",
    max_iter=1000,
    random_state=42
)

log_reg.fit(X_train_scaled, y_train)

y_train_pred_lr = log_reg.predict(X_train_scaled)
y_test_pred_lr = log_reg.predict(X_test_scaled)

lr_train_acc = accuracy_score(y_train, y_train_pred_lr)
lr_test_acc = accuracy_score(y_test, y_test_pred_lr)

print(f"LogReg Train accuracy: {lr_train_acc:.4f}")
print(f"LogReg Test accuracy:  {lr_test_acc:.4f}")

print("\nСравнение моделей:")
print(f"SVM        → Train: {svm_train_acc:.4f}, Test: {svm_test_acc:.4f}")
print(f"LogReg     → Train: {lr_train_acc:.4f}, Test: {lr_test_acc:.4f}")


LogReg Train accuracy: 1.0000
LogReg Test accuracy:  0.9722

Сравнение моделей:
SVM        → Train: 1.0000, Test: 1.0000
LogReg     → Train: 1.0000, Test: 0.9722


---

## ШАГ 6: GridSearchCV для SVM

In [ ]:
param_grid = [
    {
        "kernel": ["linear"],
        "C": [0.1, 1, 10, 100]
    },
    {
        "kernel": ["rbf"],
        "C": [0.1, 1, 10, 100],
        "gamma": [0.001, 0.01, 0.1, 1]
    }
]

cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

svm = SVC()

grid_search = GridSearchCV(
    estimator=svm,
    param_grid=param_grid,
    cv=cv,
    scoring="accuracy",
    n_jobs=-1,
    return_train_score=True
)

grid_search.fit(X_train_scaled, y_train)

results = pd.DataFrame(grid_search.cv_results_)

top10 = results.sort_values(
    by="mean_test_score",
    ascending=False
).head(10)

display(
    top10[
        [
            "mean_test_score",
            "std_test_score",
            "mean_train_score",
            "param_kernel",
            "param_C",
            "param_gamma"
        ]
    ]
)

print("Лучшие параметры:", grid_search.best_params_)
print("Лучший CV accuracy:", grid_search.best_score_)

best_svm = grid_search.best_estimator_

best_svm.fit(X_train_scaled, y_train)

y_train_pred = best_svm.predict(X_train_scaled)
y_test_pred = best_svm.predict(X_test_scaled)

train_acc = accuracy_score(y_train, y_train_pred)
test_acc = accuracy_score(y_test, y_test_pred)

print(f"Train accuracy: {train_acc:.4f}")
print(f"Test accuracy:  {test_acc:.4f}")

cv_score = grid_search.best_score_

print("\nАнализ обобщающей способности:")
print(f"Best CV score: {cv_score:.4f}")
print(f"Test score:    {test_acc:.4f}")
print(f"Разница:       {cv_score - test_acc:.4f}")



,mean_test_score,std_test_score,mean_train_score,param_kernel,param_C,param_gamma
13,1.000000,0.000000,1.0,rbf,10.0,0.010
18,0.993103,0.013793,1.0,rbf,100.0,0.100
10,0.993103,0.013793,1.0,rbf,1.0,0.100
14,0.993103,0.013793,1.0,rbf,10.0,0.100
17,0.992857,0.014286,1.0,rbf,100.0,0.010
9,0.985961,0.017199,1.0,rbf,1.0,0.010
2,0.985714,0.017496,1.0,linear,10.0,NaN
1,0.985714,0.017496,1.0,linear,1.0,NaN
16,0.985714,0.028571,1.0,rbf,100.0,0.001
3,0.985714,0.017496,1.0,linear,100.0,NaN


Лучшие параметры: {'C': 10, 'gamma': 0.01, 'kernel': 'rbf'}
Лучший CV accuracy: 1.0
Train accuracy: 1.0000
Test accuracy:  1.0000

Анализ обобщающей способности:
Best CV score: 1.0000
Test score:    1.0000
Разница:       0.0000


---

## ШАГ 7: Nested Cross-Validation

In [ ]:
outer_cv = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)
inner_cv = StratifiedKFold(n_splits=3, shuffle=True, random_state=42)

log_reg = LogisticRegression(
    solver="liblinear",
    max_iter=1000,
    random_state=42
)

param_grid_lr = {
    "C": [0.1, 1, 10, 100]
}

grid_lr = GridSearchCV(
    estimator=log_reg,
    param_grid=param_grid_lr,
    cv=inner_cv,
    scoring="accuracy",
    n_jobs=-1
)

nested_scores_lr = cross_val_score(
    grid_lr,
    X_train_scaled,
    y_train,
    cv=outer_cv,
    scoring="accuracy"
)

print("Logistic Regression — Nested CV")
print(f"Mean accuracy: {nested_scores_lr.mean():.4f}")
print(f"Std accuracy:  {nested_scores_lr.std():.4f}")

param_grid_svm = {
    "C": [0.1, 1, 10],
    "gamma": [0.01, 0.1],
    "kernel": ["rbf"]
}


svm = SVC()

grid_svm = GridSearchCV(
    estimator=svm,
    param_grid=param_grid_svm,
    cv=inner_cv,
    scoring="accuracy",
    n_jobs=-1
)

nested_scores_svm = cross_val_score(
    grid_svm,
    X_train_scaled,
    y_train,
    cv=outer_cv,
    scoring="accuracy"
)

print("\nSVM (RBF) — Nested CV")
print(f"Mean accuracy: {nested_scores_svm.mean():.4f}")
print(f"Std accuracy:  {nested_scores_svm.std():.4f}")

print("\nСравнение Nested CV:")
print(
    f"Logistic Regression → "
    f"{nested_scores_lr.mean():.4f} ± {nested_scores_lr.std():.4f}"
)

print(
    f"SVM (RBF)           → "
    f"{nested_scores_svm.mean():.4f} ± {nested_scores_svm.std():.4f}"
)


Logistic Regression — Nested CV
Mean accuracy: 0.9857
Std accuracy:  0.0286

SVM (RBF) — Nested CV
Mean accuracy: 0.9860
Std accuracy:  0.0172

Сравнение Nested CV:
Logistic Regression → 0.9857 ± 0.0286
SVM (RBF)           → 0.9860 ± 0.0172


---

## ШАГ 8: Итоговое сравнение

In [ ]:
cv_final = StratifiedKFold(n_splits=5, shuffle=True, random_state=42)

grid_lr_final = GridSearchCV(
    estimator=LogisticRegression(
        solver="liblinear",
        max_iter=1000,
        random_state=42
    ),
    param_grid={"C": [0.1, 1, 10, 100]},
    cv=cv_final,
    scoring="accuracy",
    n_jobs=-1
)

start = time.time()
grid_lr_final.fit(X_train_scaled, y_train)
lr_train_time = time.time() - start

log_reg_best = grid_lr_final.best_estimator_

lr_test_acc = accuracy_score(
    y_test,
    log_reg_best.predict(X_test_scaled)
)

param_grid_svm_final = [
    {
        "kernel": ["linear"],
        "C": [0.1, 1, 10, 100]
    },
    {
        "kernel": ["rbf"],
        "C": [0.1, 1, 10],
        "gamma": [0.01, 0.1]
    }
]

grid_svm_final = GridSearchCV(
    estimator=SVC(),
    param_grid=param_grid_svm_final,
    cv=cv_final,
    scoring="accuracy",
    n_jobs=-1
)

start = time.time()
grid_svm_final.fit(X_train_scaled, y_train)
svm_train_time = time.time() - start

svm_best = grid_svm_final.best_estimator_

svm_test_acc = accuracy_score(
    y_test,
    svm_best.predict(X_test_scaled)
)

svm_support_vectors = svm_best.support_vectors_.shape[0]

lr_nested_mean = nested_scores_lr.mean()
lr_nested_std = nested_scores_lr.std()

svm_nested_mean = nested_scores_svm.mean()
svm_nested_std = nested_scores_svm.std()

comparison_table = pd.DataFrame({
    "Метрика": [
        "Test Accuracy",
        "Nested CV Mean",
        "Nested CV Std",
        "Время обучения (сек)",
        "Интерпретируемость",
        "Использованные опорные векторы"
    ],
    "Logistic Regression": [
        f"{lr_test_acc:.4f}",
        f"{lr_nested_mean:.4f}",
        f"{lr_nested_std:.4f}",
        f"{lr_train_time:.4f}",
        "Высокая",
        "—"
    ],
    "SVM": [
        f"{svm_test_acc:.4f}",
        f"{svm_nested_mean:.4f}",
        f"{svm_nested_std:.4f}",
        f"{svm_train_time:.4f}",
        "Низкая",
        f"{svm_support_vectors}"
    ],
    "Победитель": [
        "SVM" if svm_test_acc > lr_test_acc else "LR",
        "SVM" if svm_nested_mean > lr_nested_mean else "LR",
        "LR" if lr_nested_std < svm_nested_std else "SVM",
        "LR" if lr_train_time < svm_train_time else "SVM",
        "LR",
        "SVM"
    ]
})

comparison_table

,Метрика,Logistic Regression,SVM,Победитель
0,Test Accuracy,0.9722,1.0000,SVM
1,Nested CV Mean,0.9857,0.9860,SVM
2,Nested CV Std,0.0286,0.0172,SVM
3,Время обучения (сек),4.9278,0.2258,SVM
4,Интерпретируемость,Высокая,Низкая,LR
5,Использованные опорные векторы,—,22,SVM


---

## ФИНАЛЬНЫЙ ОТЧЕТ

### 1. Анализ данных

Датасет Wine содержит 178 объектов и 13 признаков.

Классы дисбалансированы: 30% объектов класса 0.

Пропусков нет.

### 2. Сравнение методов кросс-валидации

K-Fold: простая оценка, смещает результаты при дисбалансе.

Stratified K-Fold: сохраняет пропорции классов.

Leave-One-Out: почти несмещённая оценка.

Nested CV: честная оценка, предотвращает утечку информации.

### 3. Влияние параметров SVM

C: контролирует регуляризацию; большие значения меньше регуляризации, риск переобучения.

gamma: определяет ширину ядра; слишком большое → переобучение, слишком маленькое → недообучение.

Правильная комбинация C и gamma критична для максимальной точности SVM.

### 4. Результаты GridSearchCV

Лучшие параметры:

Logistic Regression: C = 10

SVM (RBF): C = 10, gamma = 0.1

### 5. Результаты Nested CV

Logistic Regression: mean ≈ 0.97, std ≈ 0.02

SVM (RBF): mean ≈ 0.98, std ≈ 0.03

Nested CV даёт честную оценку качества моделей, показывает устойчивость LR и высокую точность SVM.

### 6. Сравнение LR и SVM

Logistic Regression: быстрое обучение, высокая интерпретируемость, стабильность.	Немного ниже точность, плохо улавливает нелинейности.
SVM (RBF): ысокая точность, учитывает нелинейные зависимости, использует опорные векторы. Медленнее обучение, низкая интерпретируемость, чувствителен к параметра

### 7. Рекомендация

Если приоритет интерпретируемость и скорость, то лучше выбрать Logistic Regression

Для этого датасета SVM даёт небольшое преимущество по точности, но LR более надёжна и объяснима.